pyqtorch/apply.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2023 pyqtorch Development Team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+from functools import reduce
+from string import ascii_letters as ABC
+from typing import Tuple
+
+import numpy as np
+import torch
+from numpy.typing import NDArray
+
+from pyqtorch.operator import Operator
+from pyqtorch.utils import reverse_permutation
+
+ABC_ARRAY: NDArray = np.array(list(ABC))
+
+
+def _apply_gate(
+    state: torch.Tensor,
+    operator: torch.Tensor,
+    qubits: list[int],
+    n_qubits: int,
+) -> torch.Tensor:
+    operator = operator.view([2] * len(qubits) * 2)
+    operator_indices = list(range(len(qubits), 2 * len(qubits)))
+    dims = (operator_indices, qubits)
+    state = torch.tensordot(operator, state, dims=dims)
+    return reverse_permutation(state, qubits, n_qubits)
+
+
+def _apply_batch_gate(
```

Comment:
Right, in hindsight I think these two functions could be merge into one `apply` by either dispatching/choosing on some parameter, possibly of `ApplyFn` type. The main point is to alleviate the definition of interdependent methods later, especially in parametric and primitive types where `apply_operator`, `apply_jacobian`, `apply_dagger`, `apply_unitary` can all be replaced by something of the sort: `apply(jacobian | dagger | operator...)` I think.

Reply:
It would be nice to do so but I am not sure how we could use single dispatch for this purpose, since I am afraid the types will be the same. Dispatching in Python is not very powerful. We could introduce an enumeration though.

Reply:
Yes indeed. There is the `ApplyFn` enum already. It can be used if dispatching is not fit for the purpose.

Reply:
dispatching will not work for the reason that @madagra pointed out.

Reply:
i would like to keep the separation though, some gates will require custom implementations of the jacobian/unitary like the PHASE gate for example

Reply:
Right, forget about dispatching it was a silly example. What I meant is that there is IMHO a proliferation of methods named `apply_*` which makes the whole code confusing. To me, it sounds much cleaner to have a single `apply` function that can be reused throughout with a meaningful signature, _i.e._ `apply(operator, state)` or `apply(matrix, state)`. Because in the end, that's the same operation used.

Reply:
sorry to hear that the code seems confusing to you. i disagree with this suggestion however, lets have a chat about it

Reply:
Cool 👍

Reply:
thanks for the suggestion and discussion @Roland-djee . i added the apply fn, please check

Reply:
circular import issues